llvm/lib/TableGen/Record.cpp

```diff
@@ -1543,6 +1543,23 @@ const Init *BinOpInit::resolveReferences(Resolver &R) const {
   const Init *lhs = LHS->resolveReferences(R);
   const Init *rhs = RHS->resolveReferences(R);
 
+  if (getOpcode() == AND) {
```

Comment:
Do we need similar short circuit for OR as well?

Reply:
Good point, I had thought this before. The problem is that since `!or` is for both logical and bitwise operation, we can't short circuit something like this `!or(true, <unknown / unresolved>)` until the second operand is resolved (in which case it's not a short circuit anymore) because we're not sure if the second operand is a int (bitwise OR) or a bit (logical OR).

That said, I think a separate logical OR / AND like you suggested might help us to apply short-circuit with more confidence.

Reply:
> Good point, I had thought this before. The problem is that since `!or` is for both logical and bitwise operation, we can't short circuit something like this `!or(true, <unknown / unresolved>)` until the second operand is resolved (in which case it's not a short circuit anymore) because we're not sure if the second operand is a int (bitwise OR) or a bit (logical OR).

I just made a blunder and forgot that `!or(<all ones>, ...)` can also be short-circuited. I'll add this case into this patch as well.

Reply:
The short-circuit for `!or` against all ones was just added.